selfservice/strategy/saml/strategy/strategy.go

```
@@ -0,0 +1,352 @@
+	return fmt.Sprintf("%s:%s", provider, subject)
+}
+
+type Strategy struct {
+	d  registrationStrategyDependencies
+	f  *fetcher.Fetcher
+	v  *validator.Validate
+	hd *decoderx.HTTP
+}
+
+func NewStrategy(d registrationStrategyDependencies) *Strategy {
+	return &Strategy{
+		d:  d,
+		f:  fetcher.NewFetcher(),
+		v:  validator.New(),
+		hd: decoderx.NewHTTP(),
+	}
+}
+
+// We indicate here that when the ACS endpoint receives a POST request, we call the handleCallback method to process it
+func (s *Strategy) setRoutes(r *x.RouterPublic) {
+	wrappedHandleCallback := strategy.IsDisabled(s.d, s.ID().String(), s.handleCallback)
+	if handle, _, _ := r.Lookup("POST", RouteAcs); handle == nil {
+		r.POST(RouteAcs, wrappedHandleCallback)
+	} // ACS SUPPORT
+}
+
+// Get possible SAML Request IDs
+func GetPossibleRequestIDs(r *http.Request, m samlsp.Middleware) []string {
+	possibleRequestIDs := []string{}
+	if m.ServiceProvider.AllowIDPInitiated {
+		possibleRequestIDs = append(possibleRequestIDs, "")
+	}
+
+	trackedRequests := m.RequestTracker.GetTrackedRequests(r)
+	for _, tr := range trackedRequests {
+		possibleRequestIDs = append(possibleRequestIDs, tr.SAMLRequestID)
+	}
+
+	return possibleRequestIDs
+}
+
+// Retrieves the user's attributes from the SAML Assertion
+func (s *Strategy) GetAttributesFromAssertion(assertion *saml.Assertion) (map[string][]string, error) {
+	attributes := map[string][]string{}
+
+	for _, attributeStatement := range assertion.AttributeStatements {
+		for _, attr := range attributeStatement.Attributes {
+			claimName := attr.Name
+			for _, value := range attr.Values {
+				attributes[claimName] = append(attributes[claimName], value.Value)
+			}
+		}
+	}
+
+	return attributes, nil
+}
+
+// Handle /selfservice/methods/saml/acs | Receive SAML response, parse the attributes and start auth flow
+func (s *Strategy) handleCallback(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	r.ParseForm()
```

Error is not checked